Web Inspector: Elements: add an editor for CSS `linear()`

dcrousso opened the pull request:
#### a933f8f850288f74e86d7d521a4106e83739719a
<pre>
Web Inspector: Elements: add an editor for CSS `linear()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=260947">https://bugs.webkit.org/show_bug.cgi?id=260947</a>

Reviewed by Brent Fulgham.

We have visual previews/editors for `cubic-bezier()`, `steps()`, and `spring()`, so we should also have something for `linear()`.

This will help developers visualize the behavior of their `linear()` function, as well as allow for better editing since there&apos;s an immediate visual feedback/preview.

* Source/WebInspectorUI/UserInterface/Views/LinearTimingFunctionEditor.js: Added.
(WI.LinearTimingFunctionEditor):
(WI.LinearTimingFunctionEditor.prototype.get element):
(WI.LinearTimingFunctionEditor.prototype.set linearTimingFunction):
(WI.LinearTimingFunctionEditor.prototype.get linearTimingFunction):
(WI.LinearTimingFunctionEditor.prototype._updateLinearTimingFunction):
(WI.LinearTimingFunctionEditor.prototype._updatePreview):
(WI.LinearTimingFunctionEditor.prototype._triggerPreviewAnimation):
(WI.LinearTimingFunctionEditor.prototype._resetPreviewAnimation):
(WI.LinearTimingFunctionEditor.prototype._updatePointsTable):
(WI.LinearTimingFunctionEditor.prototype._createPointInputs):
(WI.LinearTimingFunctionEditor.prototype._handleAddPointButtonClick):
* Source/WebInspectorUI/UserInterface/Views/LinearTimingFunctionEditor.css: Added.
(.linear-timing-function-editor):
(.linear-timing-function-editor &gt; .preview):
(.linear-timing-function-editor &gt; .preview &gt; div):
(.linear-timing-function-editor &gt; .timing):
(.linear-timing-function-editor &gt; .preview.animate &gt; div,):
(.linear-timing-function-editor &gt; svg):
(.linear-timing-function-editor &gt; svg path):
(.linear-timing-function-editor &gt; .points):
(.linear-timing-function-editor &gt; .points :is(table, thead, tbody, tfoot, tr, input)):
(.linear-timing-function-editor &gt; .points th):
(.linear-timing-function-editor &gt; .points input):

* Source/WebInspectorUI/UserInterface/Models/LinearTimingFunction.js: Added.
(WI.LinearTimingFunction):
(WI.LinearTimingFunction.fromString):
(WI.LinearTimingFunction.prototype.get points):
(WI.LinearTimingFunction.prototype.copy):
(WI.LinearTimingFunction.prototype.equals):
(WI.LinearTimingFunction.prototype.toString):
(WI.LinearTimingFunction.Point):
(WI.LinearTimingFunction.Point.prototype.get value):
(WI.LinearTimingFunction.Point.prototype.get progress):
(WI.LinearTimingFunction.Point.prototype.copy):
(WI.LinearTimingFunction.Point.prototype.equals):
* Source/WebInspectorUI/UserInterface/Models/Animation.js:
(WI.Animation.prototype._updateEffect):
* Source/WebInspectorUI/UserInterface/Views/AnimationContentView.js:
(WI.AnimationContentView.prototype._refreshPreview):
Add a new model object to represent/parse/serialize/etc. an instance of a `linear()` function.

* Source/WebInspectorUI/UserInterface/Models/TextMarker.js:
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorTextMarkers.js:
(createCodeMirrorLinearTimingFunctionTextMarkers): Added.
* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorLinearTimingFunctionEditingController.js: Added.
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.get initialValue):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.get cssClassName):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.popoverWillPresent):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.popoverDidPresent):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype._handleLinearTimingFunctionEditorLinearTimingFunctionChanged):
* Source/WebInspectorUI/UserInterface/Views/AnimationDetailsSidebarPanel.js:
(WI.AnimationDetailsSidebarPanel.prototype._refreshEffectSection):
* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor.prototype.createLinearTimingFunctionMarkers): Added.
(WI.TextEditor.prototype.editingControllerForMarker):
* Source/WebInspectorUI/UserInterface/Views/SourceCodeTextEditor.js:
(WI.SourceCodeTextEditor.prototype._updateEditableMarkers):
Add `WI.TextMarker.Type.LinearTimingFunction`.

* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js:
(WI.InlineSwatch):
(WI.InlineSwatch.prototype.didDismissPopover):
(WI.InlineSwatch.prototype._fallbackValue):
(WI.InlineSwatch.prototype._swatchElementClicked):
(WI.InlineSwatch.prototype._valueEditorValueDidChange):
* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.css:
(.inline-swatch:is(.cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .linear-timing-function, .variable)): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable)): Renamed from `.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):hover): Renamed from `.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):hover`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):active): Renamed from `.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):active`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable) &gt; span): Renamed from `.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable) &gt; span`.
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype._replaceSpecialTokens):
(WI.SpreadsheetStyleProperty.prototype._addTimingFunctionTokens):
(WI.SpreadsheetStyleProperty.prototype._addVariableTokens):
Add `WI.InlineSwatch.Type.LinearTimingFunction`.

* Source/WebInspectorUI/.eslintrc:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/animation/resources/lifecycle-utilities.js:
* LayoutTests/inspector/unit-tests/linear-timing-function.html: Added.
* LayoutTests/inspector/unit-tests/linear-timing-function-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/270855@main">https://commits.webkit.org/270855@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/771677142fc341bf88e09f82bf68337fa6549be4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16742 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18523 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15683 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20297 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17204 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17974 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17311 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14484 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14555 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15169 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19314 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15544 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15335 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21922 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19637 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15009 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6393 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
Looks nice. Can't wait to try it. r=me!

dcrousso committed:
Web Inspector: Elements: add an editor for CSS `linear()`
https://bugs.webkit.org/show_bug.cgi?id=260947

Reviewed by Brent Fulgham.

We have visual previews/editors for `cubic-bezier()`, `steps()`, and `spring()`, so we should also have something for `linear()`.

This will help developers visualize the behavior of their `linear()` function, as well as allow for better editing since there's an immediate visual feedback/preview.

* Source/WebInspectorUI/UserInterface/Views/LinearTimingFunctionEditor.js: Added.
(WI.LinearTimingFunctionEditor):
(WI.LinearTimingFunctionEditor.prototype.get element):
(WI.LinearTimingFunctionEditor.prototype.set linearTimingFunction):
(WI.LinearTimingFunctionEditor.prototype.get linearTimingFunction):
(WI.LinearTimingFunctionEditor.prototype._updateLinearTimingFunction):
(WI.LinearTimingFunctionEditor.prototype._updatePreview):
(WI.LinearTimingFunctionEditor.prototype._triggerPreviewAnimation):
(WI.LinearTimingFunctionEditor.prototype._resetPreviewAnimation):
(WI.LinearTimingFunctionEditor.prototype._updatePointsTable):
(WI.LinearTimingFunctionEditor.prototype._createPointInputs):
(WI.LinearTimingFunctionEditor.prototype._handleAddPointButtonClick):
* Source/WebInspectorUI/UserInterface/Views/LinearTimingFunctionEditor.css: Added.
(.linear-timing-function-editor):
(.linear-timing-function-editor > .preview):
(.linear-timing-function-editor > .preview > div):
(.linear-timing-function-editor > .timing):
(.linear-timing-function-editor > .preview.animate > div,):
(.linear-timing-function-editor > svg):
(.linear-timing-function-editor > svg path):
(.linear-timing-function-editor > .points):
(.linear-timing-function-editor > .points :is(table, thead, tbody, tfoot, tr, input)):
(.linear-timing-function-editor > .points th):
(.linear-timing-function-editor > .points input):

* Source/WebInspectorUI/UserInterface/Models/LinearTimingFunction.js: Added.
(WI.LinearTimingFunction):
(WI.LinearTimingFunction.fromString):
(WI.LinearTimingFunction.prototype.get points):
(WI.LinearTimingFunction.prototype.copy):
(WI.LinearTimingFunction.prototype.equals):
(WI.LinearTimingFunction.prototype.toString):
(WI.LinearTimingFunction.Point):
(WI.LinearTimingFunction.Point.prototype.get value):
(WI.LinearTimingFunction.Point.prototype.get progress):
(WI.LinearTimingFunction.Point.prototype.copy):
(WI.LinearTimingFunction.Point.prototype.equals):
* Source/WebInspectorUI/UserInterface/Models/Animation.js:
(WI.Animation.prototype._updateEffect):
* Source/WebInspectorUI/UserInterface/Views/AnimationContentView.js:
(WI.AnimationContentView.prototype._refreshPreview):
Add a new model object to represent/parse/serialize/etc. an instance of a `linear()` function.

* Source/WebInspectorUI/UserInterface/Models/TextMarker.js:
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorTextMarkers.js:
(createCodeMirrorLinearTimingFunctionTextMarkers): Added.
* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorLinearTimingFunctionEditingController.js: Added.
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.get initialValue):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.get cssClassName):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.popoverWillPresent):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype.popoverDidPresent):
(WI.CodeMirrorLinearTimingFunctionEditingController.prototype._handleLinearTimingFunctionEditorLinearTimingFunctionChanged):
* Source/WebInspectorUI/UserInterface/Views/AnimationDetailsSidebarPanel.js:
(WI.AnimationDetailsSidebarPanel.prototype._refreshEffectSection):
* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor.prototype.createLinearTimingFunctionMarkers): Added.
(WI.TextEditor.prototype.editingControllerForMarker):
* Source/WebInspectorUI/UserInterface/Views/SourceCodeTextEditor.js:
(WI.SourceCodeTextEditor.prototype._updateEditableMarkers):
Add `WI.TextMarker.Type.LinearTimingFunction`.

* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js:
(WI.InlineSwatch):
(WI.InlineSwatch.prototype.didDismissPopover):
(WI.InlineSwatch.prototype._fallbackValue):
(WI.InlineSwatch.prototype._swatchElementClicked):
(WI.InlineSwatch.prototype._valueEditorValueDidChange):
* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.css:
(.inline-swatch:is(.cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .linear-timing-function, .variable)): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable)): Renamed from `.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):hover): Renamed from `.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):hover`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):active): Renamed from `.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):active`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .linear-timing-function, .spring-timing-function, .steps-timing-function, .variable) > span): Renamed from `.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable) > span`.
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype._replaceSpecialTokens):
(WI.SpreadsheetStyleProperty.prototype._addTimingFunctionTokens):
(WI.SpreadsheetStyleProperty.prototype._addVariableTokens):
Add `WI.InlineSwatch.Type.LinearTimingFunction`.

* Source/WebInspectorUI/.eslintrc:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/animation/resources/lifecycle-utilities.js:
* LayoutTests/inspector/unit-tests/linear-timing-function.html: Added.
* LayoutTests/inspector/unit-tests/linear-timing-function-expected.txt: Added.

Canonical link: https://commits.webkit.org/270855@main
